### ISSUE TITLE
field "gene" not present in file and no sampleinfo file found

### ISSUE THREAD

**Issue by nick-youngblut** [derijkp/scywalker#10]
I'm getting the following error:

```
2024-04-18 15:45:14.641	cgjobinfo_jobid: 20748
2024-04-18 15:45:14.642	==================== sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17 ====================
2024-04-18 15:45:14.668	dependency ok (/data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/samples/barcode17/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-barcode17.tsv.zst): /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/samples/barcode17/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-barcode17.tsv.zst
2024-04-18 15:45:14.668	cgjobinfo_deps: /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/samples/barcode17/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-barcode17.tsv.zst
2024-04-18 15:45:14.669	cgjobinfo_targets: /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/samples/barcode17/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-barcode17.weighed_count.10x
2024-04-18 15:45:14.669	target missing: /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/samples/barcode17/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-barcode17.weighed_count.10x
2024-04-18 15:45:14.669 distr_submit: added to queue jobnum 8279 deps 8277 name sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17____data__SspArc0008_10x_cDNA_longRead__basecalling__scywalker__log_jobs__sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17 job /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/log_jobs/sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17 runfile /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/log_jobs/run/sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17.run options {}
   -=- starting /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker/log_jobs/sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17
   -=- 2 running 1 in queue
2024-04-18 15:45:14.682	-------------------- submitted sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-barcode17 (8279 <- 8277) (run process_project_scywalker.2024-04-18_15-28-07-504) --------------------
2024-04-18 15:45:14.683	cgjobinfo_jobid: 20749
field "gene" not present in file and no sampleinfo file found
child process exited abnormally
    while executing
"exec cg process_project  -preset scywalker -sc_whitelist $whitelist  {*}$args >@ stdout 2>@ stderr"
    (procedure "scywalker" line 9)
    invoked from within
"scywalker {*}$argv"
    (file "/usr/local/bin/scywalker-0.108.0-Linux-x86_64/apps/scywalker/scywalker.tcl" line 95)
```

My job:

```bash
docker run \
  --rm -it \
  --platform linux/amd64 \
  -v `pwd`:/data \
  --workdir /data \
  -u `id -u`:`id -g` \
  scywalker:0.108.0 \
  scywalker -v 1 -d 8 \
    -sc_expectedcells 8000 \
    -threads 6 \
    -refdir /data/references/human/scywalker/hg38 \
    /data/SspArc0008_10x_cDNA_longRead/basecalling/scywalker
```

Any idea on what the issue is?

**Comment by derijkp** [comment]
Can you rerun with the option -stack 1
That should give some extra information

**Comment by nick-youngblut** [comment]
> Can you rerun with the option -stack 1

The run with `-stack 1`:

```
2024-05-13 17:44:28.787	==================== sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2 ====================
2024-05-13 17:44:28.791	dependency ok (/large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/samples/B16_Bl6_dKO_lung_2/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2.tsv.zst): /large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/samples/B16_Bl6_dKO_lung_2/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2.tsv.zst
2024-05-13 17:44:28.792	cgjobinfo_deps: /large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/samples/B16_Bl6_dKO_lung_2/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2.tsv.zst
2024-05-13 17:44:28.792	cgjobinfo_targets: /large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/samples/B16_Bl6_dKO_lung_2/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2.weighed_count.10x
2024-05-13 17:44:28.792	target missing: /large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/samples/B16_Bl6_dKO_lung_2/sc_isoform_counts_raw-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2.weighed_count.10x
2024-05-13 17:44:28.795 distr_submit: added to queue jobnum 2116 deps 2114 name sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2____large_experiments__multiomics__SspArc0008_10x_cDNA_longRead__ont-proc_output__final__scywalker__log_jobs__sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2 job /large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/log_jobs/sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2 runfile /large_experiments/multiomics/SspArc0008_10x_cDNA_longRead/ont-proc_output/final/scywalker/log_jobs/run/sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2.run options {}
.2024-05-13 17:44:28.829	-------------------- submitted sc_filter_default_scisoformfile10x-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2 (2116 <- 2114) (run process_project_scywalker.2024-05-13_17-42-58-629) --------------------
2024-05-13 17:44:28.830	cgjobinfo_jobid: 2131
field "gene" not present in file and no sampleinfo file found
    while executing
cg select -sh /dev/null -f gene -q {$chrom eq ""} /large_experiments/multiomics/references/scywalker/hg38/extra/gene_hg38_cgencode.tsv.zst
    invoked from within
"catch_exec cg $cmd {*}$args"
    (procedure "cg" line 3)
    invoked from within
"cg select -sh /dev/null -f gene -q "\$$cfield eq \"$mitchr\"" $tsvreftranscripts"
    (procedure "sc_filter_default_job" line 60)
    invoked from within
"sc_filter_${sc_filter}_job  -reftranscripts $reftranscripts  -organelles $organelles  $scgenefile $scisoformfile $sc_expectedcells"
    (procedure "process_sample_job" line 733)
    invoked from within
"process_sample_job -todoVar todo {*}$sampleargs $dir"
    (procedure "process_project_job" line 391)
    invoked from within
"process_project_job {*}$args"
    (procedure "cg_process_project" line 9)
    invoked from within
"cg_$action {*}$args"
    invoked from within
cg process_project -preset scywalker -sc_whitelist /home/nickyoungblut/bin/scywalker-0.109.0-linux-x86_64/whitelists/3M-february-2018.txt.gz -d 8 -sc_expectedcells 8000 -threads 8 -refdir /large_experiments/multiomics/references/scywalker/hg38 scywalker

child process exited abnormally
    while executing
"exec cg process_project  -preset scywalker -sc_whitelist $whitelist  {*}$args >@ stdout 2>@ stderr"
    (procedure "scywalker" line 22)
    invoked from within
"scywalker {*}$argv"
    (file "/home/nickyoungblut/bin/scywalker-0.109.0-linux-x86_64/apps/scywalker/scywalker.tcl" line 111)
```

Note: I'm using the [hg38 reference](https://genomecomb.bioinf.be/download/refdb_hg38-0.109.0.tar.gz)

**Comment by nick-youngblut** [comment]
If I use the `g17` reference that's part of the scywalker test dataset, I do not get the error `field "gene" not present in file and no sampleinfo file found`. So, it appears that the problem is due to using the [hg38 reference](https://genomecomb.bioinf.be/download/refdb_hg38-0.109.0.tar.gz). 

However, I do get some other errors when using the `g17` reference:

* `Analysis ended with 7 errors`
* `log_jobs/sc_filter_default_r-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2`
* `log_jobs/sc_filter_write_cellinfo-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2`
* `log_jobs/sc_counts_filtered-isoquant_sc-sminimap2_splice-B16_Bl6_dKO_lung_2`
* `log_jobs/scywalker_report_B16_Bl6_dKO_lung_2`
* `samples/B16_Bl6_dKO_lung_2/log_jobs/reports_singlecell-B16_Bl6_dKO_lung_2`
* `log_jobs/reportscombine_stats-scywalker`

**Comment by derijkp** [comment]
Thank you for trying things out. I have now been able to reproduce the problem, though only with the older hg38 reference (0.106.0), not with the newer update you link to (0.109.0). 
The problem there was that sc_filter looked directly for a "gene" field instead of also trying alternate fieldnames for gene, and the transcript file of 0.106.0 did use an alternative fieldname (derijkp/genomecomb@af321b73ce5faab8c6260d4013fad069055b40e3). 

This problem does not manifest with the 0.109.0 version of hg38 for me (tested also by redownloading the one available on the website), so can you check if you actually used the latest version, or you did run with the previous (available) one.

This problem got through testing because I did not not do download testing on the reference as I do on the program (hg38 is actually the most tested reference, but the newer version used in testing was not made available on the genomecomb website), something to set up ..

**Comment by nick-youngblut** [comment]
How can one tell which version of hg38 is being used? I don't see the version listed in the `README_dbdir.txt` file, or any other file in the large collection of files in the hg38 directory.

**Comment by nick-youngblut** [comment]
With the hg38 v0.109.0 reference, I do not get the `field "gene" not present in file and no sampleinfo file found` error. However, I do get a large number of errors:

```
Analysis ended with 2774 errors
```

The errors start with the `log_jobs/map_minimap2-samples__B16_Bl6_dKO_lung_2__map-sminimap2_splice-B16_Bl6_dKO_lung_2.bam.temp__B16_Bl6_dKO_lung_2-sc.sam.zst` job.

The run log: [process_project_scywalker.2024-05-20_17-21-28-857.error.zip](https://github.com/derijkp/scywalker/files/15380262/process_project_scywalker.2024-05-20_17-21-28-857.error.zip)

**Comment by derijkp** [comment]
The minimap2 index is missing. You can create this using:
```
cg refseq_minimap2 '/large_experiments/multiomics/references/scywalker/hg38/genome_hg38.ifas' splice
```
as the error message suggests, but it is prob. easier (because creating the index requires a lot of memory) to download and unpack the premade minimap2 indexes:
```
cd /large_experiments/multiomics/references/scywalker
wget https://genomecomb.bioinf.be/download/refdb_hg38-minimap2-0.109.0.tar.gz
tar xvzf refdb_hg38-minimap2-0.109.0.tar.gz
```
I have added a warning about this (and an example) in the README

**Comment by nick-youngblut** [comment]
Thanks for helping me troubleshoot.

> The minimap2 index is missing

Maybe it would be best to have a check for necessary input at the start of the pipeline (fail fast)? Needless compute could be avoided and also the error messages could be a lot clearer. [Here](https://github.com/nf-core/rnaseq/blob/b89fac32650aacc86fcda9ee77e00612a1d77066/workflows/rnaseq.nf#L25) is an example of this approach from a popular nf-core pipeline.

**Comment by derijkp** [comment]
The error gives 2 (temp/intermediate) per-region results that have a different header. This can (afaics) only be caused by the two regions being analysed using different reftranscripts (this determines the header), though that should not happen even when continuing an interrupted analysis with a different reference (The newer reference being a dependency should cause reanalysis of the older region-results). If the error is indeed caused by a changed reference when rerunning a partially run analysis (and reanalysis of the old ones did not happen for some reason), it should be solvable by cleaning up the temp files before rerun:
```
rm -rf samples/*/isoquant_sc-*.temp
```
or (to make sure no further incompatibilities with potentially already completed parts) remove all isoquant_sc results
```
rm -rf samples/*/*isoquant_sc-*
```

> After adding the minimap2 index files to my `hg38` reference directory, I'm getting the following errors:
> 
> ```
> Analysis ended with 558 errors
> ```
> 
> The initial erroneous job is `isoquant_join-isoquant_sc-sminimap2_splice`, in which the error is:
> 
> ```
> "header of $isofile differs from header of $refisofile""\n
> ```
scywalker/files/15436494/process_project_scywalker.2024-05-23_23-57-33-710.error.zip)

**Comment by nick-youngblut** [comment]
It appears that the `field "gene" not present in file` error also occurs with the [mm10 reference](https://derijkp.github.io/genomecomb/install.html)

**Comment by derijkp** [comment]
This issue is fixed in version 0.110.0.
An updated version of mm10 is also available from the genomecomb site
